### PR TITLE
Hybrid mamba

### PR DESCRIPTION
**Pull request by fabianlim**
Drafting out the PR: There are two possibilities to support the hybrid model that was trained by `mamba_ssn` 
1. modify`mamba2` layers (i.e., blocks) to handle attention and gated MLPs. We have a `config.attention_layer_indices` that indicates the layer_ids that should be switched to attention
2. modify `jamba` layers to handle the mamba_v2 mixer. We have a `config.mamba_version` that performs this switch.

~**Items for MAMBA2**:~ : **Update: Decided to go with JAMBA**

- [x] ported over `MambaGatedMLP` from `mamba_ssn`. Currently `mamba2` blocks do not have an MLP after the mixer; this is for addressing that gap. It is equivalent to a standard GLU MLP, say, in llama. See the additional residual connection in `Mamba2Block.forward`.
- [x] ported over `MambaAttention` from `mamba_ssn`. Currently `mamba2`  blocks only have a mamba mixer; this addresses the gap of being able to switch to an attention-based mixer.
   * this results in dependency on `flash_attn`, in particular `flash_attn_with_kvcache ` and `RotaryEmbedding`.
   * for `RotaryEmbedding` allows the `rotary_dim` to be smaller that `head_dim`, which can be that case when converting a checkpoint from `mamba_ssn`.
- [x] update `Mamba2Config` to include additional configs
   * `intermediate_size`: turns on `MambaGatedMLP` if `config.intermediate_size > 0`.
   * `num_attention_heads` and `num_key_value_head` for configuring `MambaAttention`.
   * added bunch of `attention_*` flags also for configuring `MambaAttention`.
- [x] fixed bug in `Mamba2Mixer.cuda_kernels_forward` where the `ssm_state` was not saved during the first step of a generate, therefore resulting in bad generation results when the `cache` was used.
- [ ] fixed checkpoint conversion routines for converting `mamba_ssm` -> HF:
   * fix `load_state_dict_from_torch` to correctly rename the `embedding` module.
   * add logic to update `attention_*` flags if `mamba_ssm` config has `attn_cfg` setting.
   * [ ] address the bos, eos tokens, they are somehow hardcoded to all zeros, and there is no way to set these when converting `Mamba2Config` 
- [ ] refactor: remove `einops` package dependency.
- [ ] refactor: remove hardcodes on the `max_seq_len` in the cache. **Be careful of this when testing generate.**

**Items for JAMBA**:
- [x] ported over `JambaRotaryEmbedding` from mistral, which is required because `JambaAttention` does not support RoPE.
   * we ported `apply_rotary_pos_emb` and modified it such that, it can handle rotary dims that are smaller than the head dim; see comment in MAMBA2 on `RotaryEmbedding`.
   * updated `JambaAttention` and its variants to use the rotary embedding
- [x] ported over `JambaMamba2Mixer` from mamba2, to be switched via `config.mamba_version`. 
   *  currently omit the `torch_forward` method; considering if we should skip it to reduce code bloat.
- [x] update `HybridMambaAttentionDynamicCache` to handle the the different SSM cache dimensions for `config.mamba_version=="v2"`. 
- [x] update `JambaConfig` to include additional configs:
   * `mamba_version`: now supports also `"v2"` to switch to Mamba2 mixer.
   * `attn_layer_indices` in lieu of `attn_layer_period`; the latter prevents certain configurations such as the one `9, 18, 27`, it enforces `attn_layer_offset < attn_layer_period`, which is not the case in the given example.
- [x] include checkpoint conversion routine for converting a mamba2 style checkpoint from `mamba_ssm`:
   * thinking if we should leave out converting a v1 checkpoint for now.
- [x] fix `configuration` to include missing keys such as `architectures`
**Other Items**
  
There are some other items that needs to be investigated regardless of the above two paths:
- [ ] Handling `cu_seqlens` for padding free during training (scan): From the mamba_ssm repo we can see that this is [handled](https://github.com/state-spaces/mamba/blob/442fab4b1fd5226c1b5939b37d91ede430b5d1ae/mamba_ssm/modules/mamba2.py#L226) using special kernels `causal_conv1d_varlen_states` and `mamba_chunk_scan_combined`.

### CODE REVIEW

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 1]
```diff
@@ -38,7 +38,13 @@ def load_state_dict_from_safetensors(mamba2_checkpoint_path: str, ckpt_name: str
 
 
 def load_state_dict_from_torch(mamba2_checkpoint_path: str, ckpt_name: str) -> Dict[str, torch.Tensor]:
-    return torch.load(path.join(mamba2_checkpoint_path, ckpt_name), map_location="cpu")
+    original_state_dict = torch.load(path.join(mamba2_checkpoint_path, ckpt_name), map_location="cpu")
+    key_map = {
+        'backbone.embedding.weight': 'backbone.embeddings.weight',
```
Comment (fabianlim) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:43]: im wondering if this function was ever tested, because how did this rename go undetected?

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 2]
```diff
@@ -51,6 +54,12 @@
 else:
     causal_conv1d_update, causal_conv1d_fn = None, None
 
+if is_flash_attn_2_available:
+    from flash_attn import flash_attn_with_kvcache
+    from flash_attn.layers.rotary import RotaryEmbedding
+else:
+    flash_attn_with_kvcache, RotaryEmbedding = None, None
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:61]: we just stick with the mamba SSM repo and use the rotary embedding implementation from FA

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 3]
```diff
@@ -153,6 +163,11 @@ def __init__(
         self.activation = config.hidden_act
         self.act = ACT2FN[config.hidden_act]
 
+        # FIXM:E
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:166]: will need to handle the `max_seq_length` hardcodes, and also understand what `lengths_per_sample` is

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 4]
```diff
@@ -169,6 +184,214 @@ def reset(self):
...
+        self.act = ACT2FN[hidden_act]
+        self.fc2 = nn.Linear(intermediate_size, hidden_size, bias=bias)
+
+    def forward(self, hidden_states):
+        hidden_states = self.fc1(hidden_states)
+        up_proj, gate = hidden_states.chunk(2, dim=-1)
+        up_proj = up_proj * self.act(gate)
+        return self.fc2(up_proj)
+
+def _update_kv_cache(kv, cache_params: Mamba2Cache, layer_idx: int):
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:209]: can be refactored.

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 5]
```diff
@@ -359,6 +582,17 @@ def cuda_kernels_forward(
                     dim=-1,
                 )
 
+                # storing the states
+                # If we just take xBC[:, :, -self.d_conv :], it will error if seqlen < self.d_conv
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:587]: this was an existing generation bug. the `conv_states` were not properly stored in the first generation forward

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -50,6 +56,7 @@ def convert_ssm_config_to_hf_config(config_ssm: Dict, mamba2_model_dict: Dict) -
 
     # Set important values from config and recalculate other resulting entries
     hf_config.hidden_size = config_ssm[config_dict["hidden_size"]]
+    hf_config.intermediate_size = config_ssm[config_dict["intermediate_size"]]
```
Comment (fabianlim) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:59]: need to document intermediate size here. The `mixer` has the FF built inside, but this one refers to installing an additional MLP after the mixer / attn blocks

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py [hunk 7]
```diff
@@ -102,10 +125,11 @@ def load_and_save_tokenizer(
     "mamba_ssm": {
         "hidden_size": "d_model",
         "num_hidden_layers": "n_layer",
+        "intermediate_size": "d_intermediate",
         "n_groups": "ngroups",
         "bos_token_id": 0,
-        "pad_token_id": 0,
-        "eos_token_id": 0,
+        "pad_token_id": 1,
```
Comment (fabianlim) [inline src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:131]: need to understand why it was defaulted to 0 before

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 8]
```diff
@@ -169,6 +184,214 @@ def reset(self):
...
+                cache_params.lengths_per_sample
+                if cache_params.lengths_per_sample is not None
+                else cache_params.seqlen_offset
+            )
+        )
+        rotary_max_seqlen = cache_params.max_seqlen if cache_params is not None else None
+        qkv = self.in_proj(hidden_states)
+        
+        q, kv = qkv.split([self.num_heads * self.head_dim, self.num_heads_kv * 2 * self.head_dim], dim=-1)
+        q = rearrange(q, "... (h d) -> ... h d", d=self.head_dim)
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:293]: these `einops`  can be refactored out

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 9]
```diff
@@ -169,6 +184,214 @@ def reset(self):
...
+        self, 
+        hidden_states, 
+        cache_params: Optional[Mamba2Cache] = None,
+        cache_position: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+    ):
+        # if self.layer_idx == 9:
+        #     import pdb; pdb.set_trace()
+
+        if cache_params is not None and self.layer_idx not in cache_params.key_value_memory_dict:
```
Comment (fabianlim) [inline src/transformers/models/mamba2/modeling_mamba2.py:274]: The [mixer in the upstream will be able to handle](https://github.com/state-spaces/mamba/blob/442fab4b1fd5226c1b5939b37d91ede430b5d1ae/mamba_ssm/modules/mamba2.py#L256-L259) `cu_seqlens` but this MHA function does not handle it properly .


Also, this HF mixer implementation is not caught up with thte `mamba_ssn` repo and does not handle `cu_seqlens`

File: src/transformers/models/jamba/modeling_jamba.py [hunk 10]
```diff
@@ -830,6 +938,271 @@ def forward(
...
+            and seq_len == 1
+            and cache_params.conv_states[self.layer_idx].shape[0]
+            == cache_params.ssm_states[self.layer_idx].shape[0]
+            == batch_size
+        )
+
+        # getting projected states from cache if it exists
+        if use_precomputed_states:
+            in_projected_states = self.in_proj(hidden_states.squeeze(1))  # (B 2D)
+            d_mlp = (in_projected_states.shape[-1] - d_to_remove) // 2
```
Comment (fabianlim) [inline src/transformers/models/jamba/modeling_jamba.py]: consider removing this part

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 8 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/modeling_mamba2.py:293]: not doing. abandoning `modeling_mamba2`

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 7 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:131]: not doing. abandoning `modeling_mamba2`

File: src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py (hunk 6 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/convert_mamba2_ssm_checkpoint_to_pytorch.py:59]: not doing. abandoning `modeling_mamba2`

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 4 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/modeling_mamba2.py:209]: not doing. abandoning `modeling_mamba2`

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 3 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/modeling_mamba2.py:166]: not doing. abandoning `modeling_mamba2`

File: src/transformers/models/mamba2/modeling_mamba2.py (hunk 2 above)
Comment (fabianlim) [reply src/transformers/models/mamba2/modeling_mamba2.py:61]: not doing. abandoning `modeling_mamba2`
